File: Problem_1.py
"""
Problem : 1

Time Complexity : O(n)
Space Complexity : O(1)

Did this code successfully run on Leetcode : Yes
Any problem you faced while coding this : No


"""

# Task Scheduler

class Solution(object):
    def leastInterval(self, tasks, n):
        """
        :type tasks: List[str]
        :type n: int
        :rtype: int
        """
        hmap={}
        maxFreq=0
        for task in tasks:
            if task not in hmap:
                hmap[task]=0
            hmap[task]+=1
            maxFreq=max(maxFreq,hmap[task])
        maxCount=0
        for task in hmap:
            if hmap[task]==maxFreq:
                maxCount+=1
        partitions=maxFreq-1
        available=partitions*(n-maxCount+1)
        pending=len(tasks)-maxFreq*maxCount
        idle=max(0,available-pending)
        return idle+len(tasks)

File: Problem_1.js
// ## Problem1 Task Scheduler (https://leetcode.com/problems/task-scheduler/)

// Time Complexity : O(n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach

/**
 * @param {character[]} tasks
 * @param {number} n
 * @return {number}
 */
var leastInterval = function (tasks, n) {
    if (tasks === null || tasks.length === 0)
        return 0;

    let len = tasks.length;
    let map = new Map();
    let maxFreq = 0;
    // Map character and their frequency.
    // Maintain maxFreq encountered
    for (let i = 0; i < len; i++) {
        let ch = tasks[i]
        if (!map.has(ch)) {
            map.set(ch, 0);
        }
        let count = map.get(ch);
        count += 1;
        map.set(ch, count)
        maxFreq = Math.max(maxFreq, count);
    }

    // Count number of distinct chars with maxFreq
    let maxFreqCount = 0;
    map.forEach((val) => {
        if (val === maxFreq)
            maxFreqCount++;
    });
    // Tatal partitions available
    let partitions = maxFreq - 1;
    // We put chars with max freq together and then calculate remaining empty spaces
    let empty = partitions * (n - (maxFreqCount - 1));
    // Number of characters remaining to put
    let remaining = len - (maxFreq * maxFreqCount);
    // Number of Idle spaces left, these will be added to tatoal tasks to get the result
    let idle = Math.max(0, empty - remaining);

    return len + idle;
};